### ISSUE TITLE
Help with a header file

### ISSUE THREAD

**Issue by cayolblake** [ircmaxell/FFIMe#14]
Hi @bwoebi 

This isn't directly related to FFIMe however not sure how will it help. I'm still exploring how to use FFIMe after your latest fixes allowing it to at least generate the wrapper file (not sure if wrapper is the right word).

So I have the following header file, and I'm struggling with the FFI:: code that should be able to operate on the `db_prepare_request` function. I went down to working on casting and passing direct addr, however without luck and always crashes in the binary shared object of the library.

Would really, really, appreciate if you could help here.

Here's the header file.

```
#ifndef _DB_H
#define _DB_H

typedef struct db_client db_client_t;

/* Initialize the database library - must be done once per process! */
int db_init(void);

/* Client create/free */
db_client_t *db_create_client(void);
void db_free_client(db_client_t *c);

/* Prepare a request, before calling execute. If arg_lens is NULL, strings are
 * processed as null terminated. Otherwise lengths are taken from the separate
 * size_t array.
 */
int db_prepare_request(db_client_t *c, int args_count, const char **args, size_t *arg_lens);

/* Execute request; returns 0 on success. */
int db_execute(db_client_t *c);

/* Read chunk from reply. Chunking depends on internal database representation
 * so make no assumptions.
 *
 * Every call returns a pointer and updates chunk_len. When no more chunks
 * are available, NULL is returned.
 */
const char *db_read_reply_chunk(db_client_t *c, int *chunk_len);


#endif  /* _DB_H */
```

All I managed to have with no crashes is `$ffi = FFI::load("db.h"); $ffi->db_init(); $client = $ffi->db_create_client();`

Also, not sure if `FFIMe` would help making things easier right here as the examples doesn't include enough examples for various use cases for non experts.

Many thanks!

**Comment by bwoebi** [comment]
Something like `$ffi->db_prepare_request($client, 2, ["arg1", "arg2"], null);` should just work. Also, is that what you tried?

EDIT: Actually, I believe that does not work, even though it should...

If not, is it possible for you to give me a more concrete reproducer, maybe I'm generating some wrong code or such?

**Comment by cayolblake** [comment]
Thanks @bwoebi 

Yeah it doesn't work. I'll push it to a repo and send you shortly :)

**Comment by bwoebi** [comment]
I've just pushed a fix https://github.com/ircmaxell/FFIMe/commit/ffdfe3bd95e086d9abbe783b9fe850a122c85d69 - now passing strings like `["arg1", "arg2"]` should work.

**Comment by cayolblake** [comment]
@bwoebi I pushed the whole project here https://github.com/cayolblake/eredis and an already built shared object also here https://github.com/cayolblake/eredis/tree/main/build-so

**Comment by cayolblake** [comment]
@bwoebi Here's what I'm getting after update.

```
PHP Fatal error:  Uncaught Error: Call to a member function getData() on null in ~/trial1/edb.php:151
Stack trace:
#0 ~/trial1/ffime-edb.php(17): eredis\eredis->eredis_prepare_request()
#1 {main}
  thrown in ~/trial1/edb.php on line 151
```

The header file I'm pointing to is here https://github.com/cayolblake/eredis/blob/main/src/eredis.h

**Comment by cayolblake** [comment]
@bwoebi I'm calling it like `$ffi->eredis_prepare_request($c, 3, ["set", "keya", "value"], null);`

Its probably complaining because of the `null` 🤔 ?

**Comment by cayolblake** [comment]
@bwoebi My current code is.

```
<?php
require __DIR__ . '/vendor/autoload.php';
(new FFIMe\FFIMe("~/r/src/liberedis.so"))
    ->include("~/r/src/eredis.h")
    ->codeGen('eredis\\eredis', __DIR__ . '/eredis.php');
require __DIR__ . '/eredis.php';

$ffi = new eredis\eredis;
$ffi->eredis_init();
$c = $ffi->eredis_create_client(); // *** IS IT OKAY TO PASS $c TO eredis_prepare_request AS FOLLOWS? 
$ffi->eredis_prepare_request($c, 3, ["SET", "keya", "value"], null);
$ffi->eredis_execute($c);
```

**Comment by cayolblake** [comment]
@bwoebi your last commit fixed passing null issue perfectly.

Is it possible you take a look into my last comment, whether it's okay to pass client object that way?

Appreciated :)

**Comment by bwoebi** [comment]
Yes, that's totally fine. The remaining issues come from uninitialized data:
```
(gdb) p DefaultUser
$1 = (user *) 0x0
```
i.e. ACLInit() and monotonicInit() are missing.

**Comment by cayolblake** [comment]
@bwoebi Yeah, re-enabling monotonicInit() fixed some of the crashes. I checked and couldn't see ACLInit() disabled tho.

I updated the staging repo and so far no crashes. I also had another full look over the FFIMe code and can have a full execution without issues.

I can call functions without issues mainly ones that does not return a response works fine. What's missing now is getting responses.

Usage code is as follows, and I primarily have struggles with `eredis_read_reply_chunk` last parameter.

```
$ffi = new eredis\eredis;
$ffi->eredis_init();
$c = $ffi->eredis_create_client();
var_dump($ffi->eredis_prepare_request($c, 3, ["SET", "keya", "valuexxx"], null));
var_dump($ffi->eredis_execute($c));
var_dump($ffi->eredis_prepare_request($c, 2, ["GET", "keya"], null));
var_dump($ffi->eredis_execute($c));
var_dump($ffi->eredis_read_reply_chunk($c, [8]));    // **** I GUESS I'M USING IT WRONG !!
var_dump($ffi->eredis_free_client($c));
```

Also, on a side note, if it's okay, I would like to learn more how to use gdb with such setup? or, how were you able to spot the issues? I believe if I have some debugging power, things would be much easier moving forward.

**Comment by bwoebi** [comment]
Everything is fine - if you want to get the value back, then you need to pass in some variable by reference, like:
```
$chunk_len = 8;
var_dump($ffi->eredis_read_reply_chunk($c, [&$chunk_len]));
var_dump($chunk_len);
```

**Comment by bwoebi** [comment]
Using gdb is very trivially just looking like

```
root@bobtop-ubuntu:~/ffime# gdb --args php eredis_use.php
GNU gdb (Ubuntu 12.0.90-0ubuntu1) 12.0.90
[...snip...]
Reading symbols from php...
(gdb) b eredis_read_reply_chunk
Function "eredis_read_reply_chunk" not defined.
Make breakpoint pending on future shared library load? (y or [n]) y
Breakpoint 1 (eredis_read_reply_chunk) pending.
(gdb) r
Starting program: /usr/local/bin/php eredis_use.php
[Thread debugging using libthread_db enabled]
Using host libthread_db library "/lib/x86_64-linux-gnu/libthread_db.so.1".
47443:M 17 Aug 2022 16:50:27.530 * Increased maximum number of open files to 10032 (it was originally set to 1024).
47443:M 17 Aug 2022 16:50:27.530 * monotonic clock: POSIX clock_gettime
int(0)
int(0)
int(0)
int(0)

Breakpoint 1, eredis_read_reply_chunk (c=0x555556c61270, chunk_len=0x7ffff4eda640) at eredis.c:80
80      eredis.c: No such file or directory.
(gdb) p *chunk_len
$1 = 8
```

**Comment by cayolblake** [comment]
@bwoebi I did try that, however `var_dump`ing shows only the first byte of the response 🤔 

```
object(eredis\string_)#43 (1) {
  ["data":"eredis\string_":private]=>
  object(FFI\CData:char*)#40 (1) {
    [0]=>
    string(1) "+"
  }
}
int(5)
```

**Comment by bwoebi** [comment]
yes, var_dump has no idea how long the CData string is. Try actually outputting it with its length:
```
$chunk_len = 8;
$data = $ffi->eredis_read_reply_chunk($c, [&$chunk_len]);
var_dump($data->toString($chunk_len));
```

**Comment by cayolblake** [comment]
@bwoebi bloody hell... it works! :)

**Comment by cayolblake** [comment]
Its even working while setting `$chunk_len` to 0

**Comment by cayolblake** [comment]
@bwoebi it seems that it `$data->toString()` can return strings only, but not arrays.

**Comment by bwoebi** [comment]
@cayolblake I agree that it may make sense to provide a shortcut toArray() function to collect N values from a pointer. But generally the pointer classes provide ArrayAccess to get individual values.

**Comment by cayolblake** [comment]
@bwoebi Yeah it would made more sense and I guess also would be more performant.

Is it a shortcoming of PHP FFI in general?

**Comment by bwoebi** [comment]
I've just added the toArray() function.

What exactly is a shortcoming of PHP FFI?

**Comment by cayolblake** [comment]
Sorry I thought its a limitation from FFI's side. I've been reading a lot in the php-src issue tracker and can see many debates on limitations and performance issues.

Also, here's my results on using toArray() from your latest commit.

When using toString()
```
$10
helloworld
```

When using toArray()
```
array(2) {
  [0]=>
  string(1) "$"
  [1]=>
  string(1) "1"
}
```

**Comment by bwoebi** [comment]
@cayolblake Well spotted, I forgot to do a strong `null !== ` check and the `"0"` got evaluated as false too. Fixed.

**Comment by cayolblake** [comment]
@bwoebi Thanks, however sadly it crashed :)

```
TypeError: chr(): Argument #1 ($codepoint) must be of type int, string given in ~/wg1/eredis.php:240
Stack trace:
#0 ~/wg1/eredis.php(240): chr()
#1 ~/wg1/run.php(36): eredis\string_->toArray()
```

Line 240 shows ->
```
    /** @return int[] */ public function toArray(?int $length = null): array { $ret = []; if ($length === null) { $i = 0; while (null !== $cur = $this->data[$i++]) { $ret[] = \chr($cur); } } else { for ($i = 0; $i < $length; ++$i) { $ret[] = \chr($this->data[$i]); } } return $ret; }
```

**Comment by bwoebi** [comment]
@cayolblake Sorry, my bad - fixed, confused ord and chr.

**Comment by cayolblake** [comment]
@bwoebi Now its *really* crashes :)

```
126452:C 21 Aug 2022 18:33:13.130 # Accessing address: 0x561e87414000
126452:C 21 Aug 2022 18:33:13.130 # Crashed running the instruction at: 0x7fd33801ff70

------ STACK TRACE ------
EIP:
/opt/remi/php81/root/usr/lib64/php/modules/ffi.so(+0x9f70)[0x7fd33801ff70]

Backtrace:
/lib64/libc.so.6(+0x3ea70)[0x7fd337e3ea70]
/opt/remi/php81/root/usr/lib64/php/modules/ffi.so(+0x9f70)[0x7fd33801ff70]
/opt/remi/php81/root/usr/lib64/php/modules/opcache.so(+0x3f60a)[0x7fd337b3b60a]
[0x50f122a0]

------ REGISTERS ------
126452:C 21 Aug 2022 18:33:13.130 #
RAX:000000000000000e RBX:0000561e87414000
RCX:00007fd33801ff70 RDX:0000000000000000
RDI:0000561e87414000 RSI:00007fd338038ae0
RBP:00007fd337cd3300 RSP:00007ffec3656420
R8 :00007fd338038070 R9 :0000000000000000
R10:0000000000000000 R11:0000000000000246
R12:00007fd337cd35a0 R13:00007fd337c14710
R14:0000000000000000 R15:00007fd337c14740
RIP:00007fd33801ff70 EFL:0000000000010203
CSGSFS:002b000000000033
126452:C 21 Aug 2022 18:33:13.130 # (00007ffec365642f) -> 0000000184c12780
126452:C 21 Aug 2022 18:33:13.130 # (00007ffec365642e) -> 0000000042076450
126452:C 21 Aug 2022 18:33:13.130 # (00007ffec365642d) -> 0000561e84c76a05
126452:C 21 Aug 2022 18:33:13.130 # (00007ffec365642c) -> 00ff000000000000
126452:C 21 Aug 2022 18:33:13.130 # (00007ffec365642b) -> 0000000050f122a0
126452:C 21 Aug 2022 18:33:13.130 # (00007ffec365642a) -> 00007fd337c14740
126452:C 21 Aug 2022 18:33:13.130 # (00007ffec3656429) -> 0000000000000000
126452:C 21 Aug 2022 18:33:13.130 # (00007ffec3656428) -> 0000000000000000
126452:C 21 Aug 2022 18:33:13.130 # (00007ffec3656427) -> 00007fd337b3b60a
126452:C 21 Aug 2022 18:33:13.130 # (00007ffec3656426) -> 00007fd337c14680
126452:C 21 Aug 2022 18:33:13.130 # (00007ffec3656425) -> 0000000041d16c38
126452:C 21 Aug 2022 18:33:13.130 # (00007ffec3656424) -> 00007fd337c14740
126452:C 21 Aug 2022 18:33:13.130 # (00007ffec3656423) -> 00007fd337c14710
126452:C 21 Aug 2022 18:33:13.130 # (00007ffec3656422) -> 00007fd337cd35a0
126452:C 21 Aug 2022 18:33:13.130 # (00007ffec3656421) -> 00007fd337c14680
126452:C 21 Aug 2022 18:33:13.130 # (00007ffec3656420) -> 0000000041d16c38

```

**Comment by bwoebi** [comment]
I'm doing stupid things and expecting it to not crash. I need to _properly_ implement this. Will fix.

**Comment by cayolblake** [comment]
@bwoebi its okay, at least now its not crashing, but showing numerical representation of each byte,which is ord's output of each byte separately.

When using toString()
```
$10
helloworld

```

When using toArray()
```
array(17) {
  [0]=>
  int(36)
  [1]=>
  int(49)
  [2]=>
  int(48)
  [3]=>
  int(13)
  [4]=>
  int(10)
  [5]=>
  int(104)
  [6]=>
  int(101)
  [7]=>
  int(108)
  [8]=>
  int(108)
  [9]=>
  int(111)
  [10]=>
  int(119)
  [11]=>
  int(111)
  [12]=>
  int(114)
  [13]=>
  int(108)
  [14]=>
  int(100)
  [15]=>
  int(13)
  [16]=>
  int(10)
}
```

**Comment by bwoebi** [comment]
This is the expected output. toString() is giving you a string representation - toArray() gives you the representation the individual values directly.

**Comment by cayolblake** [comment]
Okay so it gives the response in byte representation? I thought (probably thinking simplicity) an array would be returned in whole words and each sub array represents new line separator.

**Comment by bwoebi** [comment]
Yeah, for your use case you should split the string returned by toString().
The ffi has no idea about the concept of line separated strings or such.

**Comment by cayolblake** [comment]
@bwoebi yeah already done that.

Is it possible to move with the address around in the code, or have to access it and bring it to PHP to be able to move it around? I mean passing the addr by reference not copying the actual data.

**Comment by bwoebi** [comment]
No, this is not possible (ffi limitation). You cannot have a direct by-ref handle to raw memory. You always have to go through explicit reading and writing of the CData handles.

**Comment by bwoebi** [comment]
What exactly do you mean? As long as you don't access the values, it won't be copied to PHP.
Once you access it (toArray(), toString(), deref() etc.), then it's copied.

Having a direct by-ref handle to raw memory is not possible though. You always have to go through explicit reading and writing of the CData handles.

Does that answer your question, or can you show me some pseudo-code of what you want to achieve?

**Comment by cayolblake** [comment]
Indeed that answers my question.

Thanks @bwoebi

**Comment by cayolblake** [comment]
@bwoebi Sorry to reopen this again, but I'm experiencing something not clear and so weird for me.

In some cases, if I do.

```
$data = $ffi->eredis_read_reply_chunk($c, [&$chunk_len]);
var_dump($data->toString());
```

I get something between 25-30 bytes long string while I should originally recive something around 200,000-250,000 bytes, yet, if I do the following...

```
$data = $ffi->eredis_read_reply_chunk($c, [&$chunk_len]);
var_dump($data->toString());
var_dump($data->toString(50));
```

In the second var_dump I get the following, which includes 30 bytes of data, different from the one before, yet, the var_dump show string(50) which is not true.

```
string(50) "(*1
+readonly
:1
:1
:1
*3
05:"
```

Then I keep increasing toString() value till I start getting more data, yet, within there are some garbage data "^�]�]A]]�". I then keep increasing more and more and around the 70,000 bytes, I start getting address access issue which might mean that I'm trying to access non existing memory addresses/blocks.

Now, what I believe is that I need to either loop over toString() with some value (which I do not know). To do that, I need to know the exact size of the data to expect.

That's all assumptions, yet, I'm sure the data is there and returned well but I'm struggling with is how to access/retrieve it all.

Thanks @bwoebi

**Comment by bwoebi** [comment]
I wonder why you are not using $chunk_len (i.e. `$data->toString($chunk_len)`), which should give you the true length - toString() is merely returning until the first NULL-byte.

If you're getting a lower $chunk_len than expected, I assume there's an issue in your usage of the library rather than in FFI.

**Comment by cayolblake** [comment]
@bwoebi I did try to use `$data->toString($chunk_len)` however `$chunk_len` is always of size 8, and if I execute `var_dump($data->toString($chunk_len));` I still get the same value every time.

It feels like there's something that should be done to move the pointer to a certain offset may be?!

**Comment by cayolblake** [comment]
Checking the C code for `eredis_read_reply_chunk`, it appears that `chunk_len` gets updated internally, yet, does not reflect in the PHP/FFI layer. No ? 🤔 

```
const char *eredis_read_reply_chunk(eredis_client_t *c, int *chunk_len)
{
    if (!c->buf_consumed) {
        listRewind(c->client->reply, &c->reply_iter);
        c->buf_consumed = 1;
        if (c->client->bufpos) {
            *chunk_len = c->client->bufpos;
            return c->client->buf;
        }
    }

    listNode *ln = listNext(&c->reply_iter);
    if (ln) {
        sds val = listNodeValue(ln);
        *chunk_len = sdslen(val);
        return val;
    } else {
        return NULL;
    }
}
```

**Comment by bwoebi** [comment]
I see no problems locally, I am _very_ sure that there is an issue with your code. Can you please share the full reproducer you have?

**Comment by cayolblake** [comment]
I will share a full reproducer code shortly, but meanwhile, I guess my issue is from PHP/FFI layer, I need to set an offset to read data from. How to do that?

**Comment by cayolblake** [comment]
As https://www.php.net/manual/en/class.ffi-cdata.php ` C array elements can be accessed as regular PHP array elements, e.g. $cdata[$offset] ` probably 🤔 !

**Comment by bwoebi** [comment]
To get a single value from a string at an offset, you can use `\chr($string[$offset])`. If you want to read a whole string easily, try using `(new string_($string->getData() + $offset))->toString($length);`

**Comment by cayolblake** [comment]
@bwoebi 

Assuming by `$string` in your examples you mean `$data` which I acquire a string from through `$data->toString()`...

I tried `var_dump( \chr($data[10]) );` and got...
```
TypeError: eredis\string_::offsetGet(): Return value must be of type eredis\str, int returned in ~/redis/eredis.php:234
```

As for the other part, `string_` is not found, not sure where to get access to it!

**Comment by bwoebi** [comment]
This is a bug in FFIMe, I'll soon look at it. With `string_` I mean the generated string_ class, in your case `eredis\string_`.

**Comment by cayolblake** [comment]
@bwoebi I did try now `$x = (new eredis\string_($data->getData() + 0))->toString(65535);` for a response that should be of size 65535 bytes and got 49131 bytes only.

**Comment by bwoebi** [comment]
Fixed the TypeError. Not sure about the size though, I would still need a reproducer here :-)

**Comment by cayolblake** [comment]
I'm prepping a repo with everything and a script to reproduce the issue clearly and consistently.

Meanwhile, I did try `var_dump( \chr($data[65535]) ); ` and got `string(1) "x"` which is a random byte from the middle of the expected string.

**Comment by cayolblake** [comment]
@bwoebi Here is the repo including everything required to reproduce.

repo: https://github.com/cayolblake/eredis/
branch: main

Steps

1. `make embedded`
2. shared object will be at `src/liberedis.so` and header at `src/eredis.h`
3. PHP/FFI script at `usage-scripts/run.php`

Appreciate it :)

**Comment by bwoebi** [comment]
I get:
```
~/eredis_new# php usage-scripts/run.php
79933:M 28 Aug 2022 23:44:22.911 * Increased maximum number of open files to 10032 (it was originally set to 1024).
79933:M 28 Aug 2022 23:44:22.911 * monotonic clock: POSIX clock_gettime

$chunk_len -------------
int(8)

$chunk_len -------------

-------------
string(8) "$65535
"
string(65551) Axxxx<many more x>xxxZ"

-------------
```

I do not see any unexpected output?

**Comment by bwoebi** [comment]
It returns 8 because that's what eredis_read_reply_chunk() actually assigns it. There's no ffi bug, but that's the behaviour of eredis.
`$65535\r\n` are 8 bytes. These are the 8 bytes assigned to chunk_len.

Now, I've been looking closer at eredis_read_reply_chunk - the `c->client->buf` contains a copy of the current buffer. `c->client->bufpos` is the amount of already consumed characters - these first 8 bytes are designating the string length.

 `((clientReplyBlock*)c->client->reply->head->value)->size` has a total amount of characters. `((clientReplyBlock*)c->client->reply->head->value)->buf` is the current position of the buffer. The remaining chunk buffer len at the current position is `((clientReplyBlock*)c->client->reply->head->value)->used`.

I don't know exactly what the proper eredis API is (where's the documentation)? But the fault is not at the FFI(Me) layer, but your usage of the eredis API being wrong.

**Comment by cayolblake** [comment]
@bwoebi Thank you for such details.

The eredis project was an attempt to create a version that is embeddable of the Redis datastore. It holds no documentation, and I had to port this 4 years old project to build against the current stable version, and with your help I was able to achieve so.

I opened yesterday an issue reporting that `eredis_read_reply_chunk` implementation isn't compatible with the current stable version, and, the maintainer disable the issue tracking functionality completely from the repo, and it was just my first issue on this project :-/

I'm no C developer at all, I used to do it 15 years ago and now more of a system engineer with a focus on well-performing solutions, yet I was able to reach this far having everything working, except for big chunked responses.

Is it possible you explain to me how you reached to know the `clientReplyBlock` struct? and it's usage? 🤔 

Is it also possible you help with point me on what to do to be able to move the buffer past the chunk that was consumed in `eredis_read_reply_chunk`? basically fixing `eredis_read_reply_chunk`!

**Comment by bwoebi** [comment]
I was just randomly inspecting the `c` struct there and its children and finding that data laying around there. I have no idea about its actual usage :-/ A documentation would be great.

I think it'd be asked a bit too much now if I had to reconstruct proper usage myself now :-|

**Comment by cayolblake** [comment]
@bwoebi yeah couldn't agree more. You've contributed a lot.

If it's possible to share your email, and you're open for a quick freelance engagement? :)

Couldn't find your email on your profile so no other way but to ask here. Sorry if it's not the right medium.

**Comment by bwoebi** [comment]
Commits include emails, when you clone a repo and check the git log :-)

And, no, I'm not available for that. I'll gladly help you with issues on ffime, but that's just whenever I get a little time during the day, I don't want to commit much of my time though.

**Comment by cayolblake** [comment]
@bwoebi Yeah sure, probably too much pressure I didn't think of the trivial things like checking commit log :)

To wrap this up, I appreciate your involvement in FFIMe, also keeping up that long! 👍
